clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -217,104 +235,146 @@ private boolean flush(ByteBuffer buf) throws IOException {
     * | unwrap()    | Finished                         | FINISHED    |
     * +-------------+----------------------------------+-------------+
     *
-    * @throws IOException
+    * @throws IOException if read/write fails
+    * @throws SslAuthenticationException if handshake fails with an {@link SSLException}
     */
     @Override
     public void handshake() throws IOException {
+        int read = 0;
+        // Read any available bytes before attempting any writes to ensure that handshake failures
+        // reported by the peer are processed even if writes fail (since peer closes connection
+        // if handshake fails)
+        if (key.isReadable())
+            read = readFromSocketChannel();
+
+        try {
+            doHandshake();
+        } catch (SSLException e) {
+            handleSslException(e, true);
```

Comment:
One thing that wasn't exactly clear to me is what happens if flush returns false. We won't throw the exception here, but we could potentially throw it in the `read == -1` check below. Also, in the next iteration, we seem to overwrite the handshake status. Shouldn't we be checking if the handshake already failed? Or am I missing something?

Reply:
Oops, I had removed the overrwrite of handshake status earlier, but put it back yesterday without thinking. Have updated the code.